### PR TITLE
Add tool to visualize operations 

### PR DESCRIPTION
**Pull request by bmerry**
It shows the hierarchy of operations and slots. Graphviz is required.

### CODE REVIEW

**Comment by bmerry** [comment]
Here's an example of the output: [ingest.gv.pdf](https://github.com/ska-sa/katsdpsigproc/files/3555393/ingest.gv.pdf)

File: katsdpsigproc/accel.py [hunk 1]
```diff
@@ -1436,3 +1437,116 @@ def _extract_slots(self, names, add_to_hidden):
...
+    operation : :class:`Operation`
+        Operation to add to the graph
+    path : tuple of str
+        Hierarchical name of `operation`. The last component is used for visual
+        display, and the rest for generating unique node labels.
+    slot_map : dict
+        Mapping from slots to node names (or port names). It is updated in place
+        with all the slots in this operation and its descendants.
+    no_detail_slots : set
+        Slots would should not show shape and dtype details. It is updated in
```
Comment (ludwigschwardt) [inline katsdpsigproc/accel.py]: would should

File: katsdpsigproc/accel.py [hunk 2]
```diff
@@ -1436,3 +1437,116 @@ def _extract_slots(self, names, add_to_hidden):
...
+                    # ":" has special meaning to graphviz, so we replace it with "+".
+                    full_name = 'slot_{}_{}'.format('!'.join(path), slot_name.replace(':', '+'))
+                    slot_map[slot] = full_name
+                    label = '<' + _slot_label(slot, slot_name, slot in no_detail_slots) + '>'
+                    sub.node(full_name, label=label, shape='box')
+                else:
+                    del all_slots[slot_name]
+    else:
+        node_name = 'op_' + '!'.join(path)
+        label = io.StringIO()
```
Comment (ludwigschwardt) [inline katsdpsigproc/accel.py]: Why `label = io.StringIO(); label.write('...')` and not `label = ''; label += '...'`? (I guess that's one of  StringIO's strengths? Automatic line termination?)

File: katsdpsigproc/accel.py [hunk 3]
```diff
@@ -1436,3 +1437,116 @@ def _extract_slots(self, names, add_to_hidden):
...
+    """Write a visualization of an :class:`Operation` to file.
+
+    This requires :mod:`graphviz` to be installed.
+
+    Parameters
+    ----------
+    operation : :class:`Operation`
+        Operation or operation sequence to visualize
+    filename : str
+        Base filename to write. It should end in ``.gv``, and a rendered
```
Comment (ludwigschwardt) [inline katsdpsigproc/accel.py]: Is the `.gv` file stored too?

**Comment by ludwigschwardt** [comment]
Nice!

File: katsdpsigproc/accel.py (hunk 3 above)
Comment (bmerry) [reply katsdpsigproc/accel.py]: Yes.

File: katsdpsigproc/accel.py (hunk 2 above)
Comment (bmerry) [reply katsdpsigproc/accel.py]: Building a string by repeated concatenation can potentially take quadratic time, so is generally an anti-pattern (although CPython has some optimisations for it). Using StringIO gives linear-time performance.

Comment (bmerry) [reply katsdpsigproc/accel.py]: Another option would be to collect the pieces and use `''.join` at the end. That might be a little more natural - what do you think?

Comment (ludwigschwardt) [reply katsdpsigproc/accel.py]: That makes sense! No preference - we can leave it as is.

File: katsdpsigproc/accel.py (hunk 1 above)
Comment (bmerry) [reply katsdpsigproc/accel.py]: Fixed.

**Pull request merged**